trie/triedb/pathdb/disklayer.go

```diff
@@ -0,0 +1,259 @@
+	}
+	dl.stale = true
+}
+
+// Node retrieves the trie node with the provided node info. No error will be
+// returned if the node is not found.
+func (dl *diskLayer) Node(owner common.Hash, path []byte, hash common.Hash) ([]byte, error) {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	if dl.stale {
+		return nil, errSnapshotStale
+	}
+	// Try to retrieve the trie node from the not-yet-written
+	// node buffer first. Note the buffer is lock free since
+	// it's impossible to mutate the buffer before tagging the
+	// layer as stale.
+	n, err := dl.buffer.node(owner, path, hash)
+	if err != nil {
+		return nil, err
+	}
+	if n != nil {
+		dirtyHitMeter.Mark(1)
+		dirtyReadMeter.Mark(int64(len(n.Blob)))
+		return n.Blob, nil
+	}
+	dirtyMissMeter.Mark(1)
+
+	// Try to retrieve the trie node from the clean memory cache
+	if dl.db.cleans != nil {
+		if blob := dl.db.cleans.Get(nil, hash.Bytes()); len(blob) > 0 {
+			cleanHitMeter.Mark(1)
+			cleanReadMeter.Mark(int64(len(blob)))
+			return blob, nil
+		}
+		cleanMissMeter.Mark(1)
+	}
+	// Try to retrieve the trie node from the disk.
+	var (
+		nBlob []byte
+		nHash common.Hash
+	)
+	if owner == (common.Hash{}) {
+		nBlob, nHash = rawdb.ReadAccountTrieNode(dl.db.diskdb, path)
+	} else {
+		nBlob, nHash = rawdb.ReadStorageTrieNode(dl.db.diskdb, owner, path)
+	}
+	if nHash != hash {
+		return nil, &UnexpectedNodeError{
+			typ:      "disk",
+			expected: hash,
+			hash:     nHash,
+			owner:    owner,
+			path:     path,
+		}
+	}
+	if dl.db.cleans != nil && len(nBlob) > 0 {
+		dl.db.cleans.Set(hash.Bytes(), nBlob)
+		cleanWriteMeter.Mark(int64(len(nBlob)))
+	}
+	return nBlob, nil
+}
+
+// update returns a new diff layer on top with the given state set.
+func (dl *diskLayer) update(stateRoot common.Hash, id uint64, blockNumber uint64, nodes map[common.Hash]map[string]*trienode.Node, states *triestate.Set) *diffLayer {
+	return newDiffLayer(dl, stateRoot, id, blockNumber, nodes, states)
+}
+
+// commit merges the given bottom-most diff layer into the node buffer
+// and returns a newly constructed disk layer. Note the current disk
+// layer must be tagged as stale first to prevent re-access.
+func (dl *diskLayer) commit(bottom *diffLayer, force bool) (*diskLayer, error) {
+	dl.lock.Lock()
+	defer dl.lock.Unlock()
+
+	// Construct and store the state history first. If crash happens
+	// after storing the state history but without flushing the
+	// corresponding states(journal), the stored state history will
+	// be truncated in the next restart.
+	if dl.db.freezer != nil {
+		err := writeStateHistory(dl.db.freezer, bottom, dl.db.config.StateLimit)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Mark the diskLayer as stale before applying any mutations on top.
+	dl.stale = true
+
+	// Store the root->id lookup afterwards. All stored lookups are
+	// identified by the **unique** state root. It's impossible that
+	// in the same chain blocks are not adjacent but have the same
+	// root.
```

Comment:
nonce increment. If the state transition is not empty(with transactions included), the tx senders' nonce must be bumped.

at least this fact is still held in current spec, but in the future, for example, we can do some state transitions without any transaction(e.g. system call or whatever) and somehow the state can be reverted in the next block, then it will be problematic. But I think we should prevent this thing in the first place.